### PR TITLE
Rust bindings API review

### PR DESCRIPTION
**Pull request by rafaelcaricio**
This PR makes the Rust bindings usage more ergonomic and flexible.

### CODE REVIEW

File: lvgl/src/lv_core/obj.rs [hunk 1]
```diff
@@ -176,6 +177,38 @@ macro_rules! define_object {
...
+//
+//     pub fn create_at(parent: &mut impl crate::NativeObject) -> crate::LvResult<Self> {
+//         Ok(Self::create(parent, None)?)
+//     }
+//
+//     pub fn new() -> crate::LvResult<Self> {
+//         let mut parent = crate::display::DefaultDisplay::get_scr_act()?;
+//         Ok(Self::create_at(&mut parent)?)
+//     }
+// }
```
Comment (rafaelcaricio) [inline lvgl/src/lv_core/obj.rs:210]: This is used during development, to test the code generation options.

File: lvgl/src/mem.rs [hunk 2]
```diff
@@ -69,14 +69,33 @@ impl<T> AsMut<T> for Box<T> {
     }
 }
 
+fn mem_info() -> lvgl_sys::lv_mem_monitor_t {
```
Comment (rafaelcaricio) [inline lvgl/src/mem.rs:72]: Probably just need to make safe bindings for this simple function and add to the public API. It's useful for app debugging.

File: lvgl/src/widgets/label.rs [hunk 3]
```diff
@@ -10,6 +13,16 @@ impl Label {
     }
 }
 
+#[cfg(feature = "alloc")]
+impl<S: AsRef<str>> From<S> for Label {
+    fn from(text: S) -> Self {
+        let text_cstr = CString::new(text.as_ref()).unwrap();
```
Comment (rafaelcaricio) [inline lvgl/src/widgets/label.rs:19]: This is a test of what is possible with `alloc` feature enabled. Actualy the `From` trait implementation should use the `TryFrom` trait implementation and just call `.unwrap()` here. People will decide which guarantees they want to have.

File: lvgl/src/lib.rs [hunk 4]
```diff
@@ -47,22 +44,60 @@ pub(crate) mod mem;
...
+
+#[cfg(test)]
+pub(crate) mod tests {
+    use super::*;
+    use crate::display::{Display, DrawBuffer};
+    use embedded_graphics::mock_display::MockDisplay;
+    use embedded_graphics::pixelcolor::Rgb565;
+
+    pub(crate) fn initialize_test() {
+        init();
```
Comment (rafaelcaricio) [inline lvgl/src/lib.rs:91]: All tests run in the same LVGL "instance" so the tests might run out of memory at some point. :) Just to keep in mind.

File: lvgl/src/display.rs [hunk 5]
```diff
@@ -0,0 +1,359 @@
...
+            .enumerate()
+            .map(|(iy, y)| {
+                xs.clone().map(move |(ix, x)| {
+                    let color_len = x_len * iy + ix;
+                    let lv_color = unsafe { *color_p.add(color_len) };
+                    let raw_color = Color::from_raw(lv_color);
+                    drawable::Pixel::<C>(Point::new(x as i32, y as i32), raw_color.into())
+                })
+            })
+            .flatten();
```
Comment (rafaelcaricio) [inline lvgl/src/display.rs:302]: TODO: Remove this code duplication.

File: lvgl/src/display.rs [hunk 6]
```diff
@@ -0,0 +1,359 @@
...
+impl DefaultDisplay {
+    /// Gets the screen active of the default display.
+    pub fn get_scr_act() -> Result<Obj> {
+        Ok(get_str_act(None)?)
+    }
+}
+
+pub struct DrawBuffer<const N: usize> {
+    initialized: RunOnce,
+    refresh_buffer: Mutex<RefCell<heapless::Vec<lvgl_sys::lv_color_t, N>>>,
```
Comment (rafaelcaricio) [inline lvgl/src/display.rs:91]: I use a `Mutex` here, to satisfy the Rust guarantees. But maybe it is not necessary... some food for thought.

File: lvgl-sys/build.rs [hunk 7]
```diff
@@ -114,6 +114,7 @@ fn main() {
     let bindings = bindgen::Builder::default()
         .header(shims_dir.join("lvgl_sys.h").to_str().unwrap())
         .generate_comments(false)
+        .derive_default(true)
```
Comment (rafaelcaricio) [inline lvgl-sys/build.rs:117]: Extremely useful feature of `bindgen`. 💅🏽

File: lvgl/src/display.rs [hunk 8]
```diff
@@ -0,0 +1,268 @@
...
+impl DefaultDisplay {
+    /// Gets the screen active of the default display.
+    pub fn get_scr_act() -> Result<Obj> {
+        Ok(get_str_act(None)?)
+    }
+}
+
+pub struct DrawBuffer<const N: usize> {
+    initialized: RunOnce,
+    refresh_buffer: Mutex<RefCell<[MaybeUninit<lvgl_sys::lv_color_t>; N]>>,
```
Comment (rafaelcaricio) [inline lvgl/src/display.rs:67]: I'm using a Mutex here just to make this `Sync`, not sure if there is a way to remove it but still keep the memory statically allocated.

File: examples/app.rs [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+    let output_settings = OutputSettingsBuilder::new().scale(2).build();
+    let mut window = Window::new("App Example", &output_settings);
+
+    let mut shared_native_display = Arc::new(Mutex::new(embedded_graphics_display));
+
+    // LVGL usage
+    lvgl::init();
+
+    const REFRESH_BUFFER_SIZE: usize = lvgl::DISP_HOR_RES * lvgl::DISP_VER_RES / 10;
+    static DRAW_BUFFER: DrawBuffer<REFRESH_BUFFER_SIZE> = DrawBuffer::new();
```
Comment (rafaelcaricio) [inline examples/app.rs:28]: Making this statically allocated will avoid dropping this memory address by mistake in a app code.

**Comment by nia-e** [comment]
What's the status on this PR? I was considering attempting to upgrade LVGL to 8.x and it seems like this is necessary. I can rebase on current master and flesh out any missing features if that would be desirable.

**Comment by nia-e** [comment]
To my utter surprise, a rebase and minimal fixes later, this compiles. I'd rather test more things before I do anything but it seems like my job is easier than expected. Should I open a new PR or push to this branch? I squashed all commits.

**Comment by rafaelcaricio** [comment]
@nia-e Awesome, Rust is great. lol

You can commit to this branch and continue this PR.

**Comment by nia-e** [comment]
What were the plans for `lvgl_codegen`? It has some unused code and behavior is missing. I could try to finish implementing whatever the original idea was.

**Comment by nia-e** [comment]
Also; further digging on the `Mutex`es revealed that there is Probably No Way to get rid of them short of marking this entire library as non-thread-safe which doesn't feel great. I'd love to be proven wrong, though. `parking_lot` is a very small crate though and this shouldn't be a huge deal - certainly far better than pulling in all of `embedded-graphics` at all times like before.

**Comment by rafaelcaricio** [comment]
@nia-e The idea with the `lvgl-codegen` crate was to be able to generate the entirety of the LVGL C API instead of manually writing bindings for every single API function LVGL exposes. I wanted to have in the `lvgl` crate the "core abstractions" that would be used by the codegen. This was [a suggestion](https://forum.lvgl.io/t/rust-bindings-for-littlevgl/2386/9?u=ioctl) I got at the begging of this project. That thread might be a good source of what I was thinking at the time too (it has been a while ago 😬 ).

[LVGL is not thread-safe itself](https://docs.lvgl.io/master/porting/os.html), so making this crate non-thread-safe shouldn't be a bid deal? 👀 For us to make this crate "thread-safe" we would need to (have and) lock a global `Mutex` every time we call LVGL C API, not sure if that is a good idea!?

**Comment by nia-e** [comment]
Perfect, okay. I'll drop the `Mutex`es and read through that thread afterwards.

**Comment by nia-e** [comment]
Seems like extending `lvgl-codegen` is a can I can safely kick down the road. All in all, I think this is complete and ready for review. Squashed commits.

**Review by nia-e — APPROVED** [review]

**Comment by nia-e** [comment]
Okay, now it's ready. Give this a look when you have some time. Also this hopefully beings to close #42

**Review by nia-e — APPROVED** [review]

**Comment by rafaelcaricio** [comment]
LGTM. I cannot approve as I originally opened the PR. Feel free to merge ;)

**Pull request merged**